### PR TITLE
Merge daemon defaults with user-supplied jvmargs

### LINKED ISSUE #19750: Setting `org.gradle.jvmargs` property can lead to "Daemon disappeared" failures
When investigating reports of "daemon disappeared" in GitHub Actions, and [a feature request to address one cause](https://github.com/gradle/gradle-build-action/issues/122), it was found that the all of our DEFAULT_JVM8_ARGS are being lost as soon as _any_ value is set for `org.gradle.jvmargs`.

When `org.gradle.jvmargs` is not set by the user, then the daemon will start with default values [specified here](https://github.com/gradle/gradle/blob/6ccf8a2cc74c8060992bd36189dec929143caed7/subprojects/launcher/src/main/java/org/gradle/launcher/daemon/configuration/DaemonParameters.java#L38-L39). But setting a value of `org.gradle.jvmargs` will overwrite these defaults, resulting in JDK defaults for `-Xmx`, `-Xms` etc.

One of the key defaults set by Gradle is `-XX:MaxMetaspaceSize=256m`, which is required because [this value is unbounded on the JDK by default](https://docs.oracle.com/en/java/javase/17/gctuning/other-considerations.html#GUID-B29C9153-3530-4C15-9154-E74F44E3DAD9), allowing the Gradle daemon to consume more and more native memory until it crashes. 

This means that a user setting `org.gradle.jvmargs=-Xmx4g` can result in daemon crashes, unless the user knows that they must _also_ add a `-XX:MaxMetaspaceSize` setting.

In a similar way, the `-XX:+HeapDumpOnOutOfMemoryError` default is lost when a user specifies `org.gradle.jvmargs`.

### Expected Behavior

Important JVM defaults are not lost when a user sets unrelated JVM arguments.

### Current Behavior

Users must be aware of all important JVM argument defaults, and include all of these when setting any values for `org.gradle.jvmargs`.

### Context

This was discovered when investigating "daemon disappeared" issues in GitHub Actions workflows.
Here is an example of a build with such a failure: https://github.com/androidx/androidx/runs/4971463356?check_suite_focus=true#step:10:3845. No build scan was captured due to the nature of the failure.

Daemon log file for that CI run:
[gradle-daemon-logs_room.zip](https://github.com/gradle/gradle/files/7958782/gradle-daemon-logs_room.zip)

### PR DESCRIPTION
**Pull request by bigdaz**
Fixes #19750

### Context

Previously, setting any value for `org.gradle.jvmargs` caused all default settings to be lost.
This often resulted in important defaults like `-XXMaxMetaspaceSize` being omitted when a user
attempted to provide more memory to a build process.

With this change, default jvmargs will be retained unless specifically overridden by a user-supplied argument.
One exception is that setting either -Xmx or -Xms will cause the default heap size settings to be omitted, preventing
user-supplied values from conflicting with default values (like having a min heap larger than max heap).

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by bigdaz** [comment]
@bot-gradle test this

**Comment by bigdaz** [comment]
@bot-gradle test this

**Comment by bigdaz** [comment]
@bot-gradle test this

**Comment by bigdaz** [comment]
@bot-gradle test this

**Comment by bigdaz** [comment]
@bot-gradle test this

**Comment by bigdaz** [comment]
@bot-gradle test this

**Review by octylFractal — CHANGES_REQUESTED** [review]
Thanks for your contribution. I think the change looks good overall, but I would like some more docs covering the new merging behavior.

**Comment by bigdaz** [comment]
@octylFractal Thanks for the feedback. Unfortunately this PR is hitting some [hard-to-understand test failures](https://ge.gradle.org/s/tpud6st67opl4/tests/overview?outcome=failed) that will require further investigation before this change could be merged. I'm not sure if/when I'll find time to do that.

In the meantime, would it be helpful if I close this PR (or convert to draft?)

**Comment by octylFractal** [comment]
I'm looking at the test failures now, so don't worry about that.

File: subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc [hunk 1]
```diff
@@ -96,7 +96,7 @@ _A reasonable default is derived from your environment (`JAVA_HOME` or the path
 
 `org.gradle.jvmargs=(JVM arguments)`::
 Specifies the JVM arguments used for the Gradle Daemon. The setting is particularly useful for <<#sec:configuring_jvm_memory,configuring JVM memory settings>> for build performance. This does not affect the JVM settings for the Gradle client VM.
```
[the remark this answers is not in the document]
Comment (bigdaz) [reply subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc]: I've pushed some improvements to the doc. PTAL: 438a568

[the remark this answers is not in the document]
Comment (octylFractal) [reply subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc]: I made the wording better (in my opinion), PTAL: 24a222e8a36f751b598c0f60150af0d27bf9c008

**Comment by octylFractal** [comment]
@bot-gradle test this

**Comment by octylFractal** [comment]
@bigdaz I looked into the "hard-to-understand test failures", and the underlying issue is that we used to let anything that set `org.gradle.jvmargs` run with unlimited `Metaspace` if not specified, and now we limit it to `256m`. After some research, @DPUkyle and I came to the conclusion that it would be a good idea to increase this to perhaps `1G` by default, or remove it entirely. What do you think?

**Comment by bigdaz** [comment]
We should not remove the setting altogether. The entire reason for this PR is that users are setting `org.gradle.jvmargs` _without_ specifying `MaxMetaspaceSize`, and the daemon process is consuming more and more memory until the process dies.
See #19750 for details.

Providing a higher default value might make sense. Or since these failures seem specific to the Kotlin compiler daemon (I think), perhaps the best fix is to ensure an appropriate `MaxMetaspaceSize` for this process.

**Comment by bigdaz** [comment]
@eskatos I see you assigned this back to me, but I'm not sure what action can/should be taken.
The current behaviour is that if you don't set _any_ jvm args, then we set `MaxMetaspaceSize=256m`. If a user sets a completely unrelated jvm arg, then we don't provide any default value for `MaxMetaspaceSize`.

Perhaps this makes sense: we provide a default _set_ of jvm args and you can choose to override the entire set, but not one value in the set. But I've seen a number of users struggle with this: they give their build _more_ memory (say `-Xmx1024m`), and their build starts to run out of memory! It's a slightly different out of memory error, but users don't really know how to fix this.

There are a few options to address this:
1. Do nothing, but perhaps improve the documentation so it's clear that setting `-Xmx` does more than just giving more memory.
2. When a user sets just one of the values that we provide a default for, we only change _that_ value, and leave the others in place. (That's what my PR tries to do).
3. Provide a different (or higher) default value for `MaxMetaspaceSize`. We should probably document this. 
4. Provide a better model for specifying the memory settings for a build, allowing users to set value independently.
5. ???

I could help out with any of 1-3. I think 4 would require more work and should be tackled by the BT team.
WDYT?

**Comment by eskatos** [comment]
I assigned it back to you while triaging unassigned PRs and adding assignees to all team members/authors.

I think this PR is a reasonable improvement and will address the current confusion. But as shown by the failing tests, this may break builds.

I'm not sure if raising the default metaspace size to 1g would be a good move. Loots of builds won't need that much.

@octylFractal, @big-guy, would the approaching 8.0 be a good time to address this? Can you take over making a decision on this?

**Comment by jvandort** [comment]
@bot-gradle test ReadyForNightly

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
The performance test asserts there's only one daemon involved in multiple iterations. However, this PR seems to change something in daemon compatibility so there're multiple daemon started. The performance test failure is reproducbile.

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc [hunk 2]
```diff
@@ -46,7 +46,13 @@ TBD
...
 === Potential breaking changes
 
-TBD
+==== User provided daemon JVM arguments are merged with defaults
+
+Prior to 8.0, setting `org.gradle.jvmargs` would overwrite all default daemon JVM args, however now Gradle's default JVM arguments are merged with the arguments provided by this property. In some cases, this may cause builds which set `org.gradle.jvmargs` to have different memory configuration in Gradle 7.6.
+
+As it was before, the default daemon JVM args are `-Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError`. Previously, if `org.gradle.jvmargs` is set to `-DsomeProperty=123`, the existing behavior would be to resolve the daemon arguments to `-DsomeProperty=123` while the new behavior would resolve to `-DsomeProperty=123 -Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError`.
+
+Since the JVM's default memory arguments are more generous than Gradle's default daemon JVM arguments, this change can cause existing builds to fail with an `OutOfMemoryError` if the build uses more memory than is explicitly set in `org.gradle.jvmargs`.
```
Comment (nathan-contino) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc:55]:
```suggestion
==== User provided daemon JVM arguments merge with defaults

Prior to 8.0, setting `org.gradle.jvmargs` overwrites all default daemon JVM args. In 8.0 and above, Gradle merges default JVM arguments with the arguments provided by this property. This can effect the memory configuration of builds that set `org.gradle.jvmargs`.

Consider the default daemon JVM arguments:

----
-Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError
----

Prior to 8.0, if a user assigned `org.gradle.jvmargs` the value `-DsomeProperty=123`, Gradle resolved the daemon arguments to the following:

----
-DsomeProperty=123
----

In 8.0 and above, if a user assigns `org.gradle.jvmargs` the value `-DsomeProperty=123`, Gradle resolves the daemon arguments to the following:

----
-DsomeProperty=123 -Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError`
----

Since the JVM's default memory arguments are more generous than Gradle's default daemon JVM arguments, this change can cause existing builds to fail with an `OutOfMemoryError` if the build uses more memory than explicitly set in `org.gradle.jvmargs`.
```

Suggestion: use the terms "Before 8.0" and "In 8.0 and above" to make the subject of sentences (and comparisons) clearer.

Separate large collections of JVM arguments into block code snippets to make them more readable.

Use active voice when possible to make it clearer when Gradle behaves in some way, and when the user behaves in some way. I find the docs a little hard to parse as written.

File: subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc [hunk 3]
```diff
@@ -100,8 +100,9 @@ Specifies the Java home for the Gradle build process. The value can be set to ei
 _A reasonable default is derived from your environment (`JAVA_HOME` or the path to `java`) if the setting is unspecified._
 
 `org.gradle.jvmargs=(JVM arguments)`::
-Specifies the JVM arguments used for the Gradle Daemon. The setting is particularly useful for <<#sec:configuring_jvm_memory,configuring JVM memory settings>> for build performance. This does not affect the JVM settings for the Gradle client VM.
-_The default is `-Xmx512m "-XX:MaxMetaspaceSize=256m"`._
+Specifies JVM arguments to be used for the Gradle Daemon. The setting is particularly useful for <<#sec:configuring_jvm_memory,configuring JVM memory settings>> for build performance. This does not affect the JVM settings for the Gradle client VM.
+Supplied values are merged with the defaults, with special handling for <<#sec:configuring_jvm_memory,JVM memory settings>> (see link).
+_The default values are `-Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError`._
```
Comment (nathan-contino) [inline subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc:105]:
```suggestion
Specifies the JVM arguments to use for the Gradle Daemon. The setting is particularly useful for <<#sec:configuring_jvm_memory,configuring JVM memory settings>> for build performance. This does not affect the JVM settings for the Gradle client VM.
Gradle merges supplied values with the defaults, with special handling for <<#sec:configuring_jvm_memory,JVM memory settings>>.
_The default value is `-Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError`._
```
Suggestion: slight reword for docs consistency. I suspect we should specify the default *value* here (not values) because the default value refers to the singular `org.gradle.jvmargs` property, not the plural JVM arguments.

File: subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc [hunk 4]
```diff
@@ -325,11 +326,14 @@ You should check for existence of optional project properties before you access
 
 You can adjust JVM options for Gradle in the following ways:
 
-The `org.gradle.jvmargs` Gradle property controls the VM running the build. It defaults to `-Xmx512m "-XX:MaxMetaspaceSize=256m"`
+The `org.gradle.jvmargs` Gradle property specifies arguments for the JVM running the build. The default values are `-Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError`.
```
Comment (nathan-contino) [inline subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc:329]:
```suggestion
The `org.gradle.jvmargs` Gradle property specifies arguments for the JVM running the build. The default value is `-Xms256m -Xmx512m -XX:MaxMetaspaceSize=256m -XX:+HeapDumpOnOutOfMemoryError`.
```
Same suggestion as above -- I *think* this ought to be singular.

File: subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc [hunk 5]
```diff
@@ -325,11 +326,14 @@ You should check for existence of optional project properties before you access
...
 
-.Changing JVM settings for the build VM
+JVM arguments provided via `org.gradle.jvmargs` are combined with the defaults, with special handling for `-Xms` and `-Xmx`: setting either these values will remove the default for the other.
+This avoids the case where a default value is incompatible with a user-supplied value, such as having a minimum memory setting higher than the maximum.
```
Comment (nathan-contino) [inline subprojects/docs/src/docs/userguide/running-builds/build_environment.adoc:332]:
```suggestion
Gradle combines JVM arguments provided via `org.gradle.jvmargs` with the default JVM arguments. Setting either `-Xms` and `-Xmx` removes the default for the other.
This avoids incompatible combinations of default and user-supplied values, such as a minimum memory setting higher than the maximum.
```
Suggestion: active voice rephrase to make the first sentence more readable. Trimmed some words from the second sentence for the same reason.

**Review by nathan-contino — APPROVED** [review]
Docs LGTM, a couple of optional suggestions

**Comment by big-guy** [comment]
We're not able to look at this until 2025

**Comment by runningcode** [comment]
Bump. This leads to so much frustration with Gradle users who aren’t aware of this footgun.

**Comment by bigdaz** [comment]
@big-guy I understand that this was considered a breaking change, but can we please get the behaviour changed for Gradle 9.0?

**Comment by 3flex** [comment]
This best practice doesn't mention this problem: https://docs.gradle.org/9.2.0-rc-1/userguide/common_caching_problems.html#system_file_encoding

Breaking changes were mentioned above, so is this deferred to at least 10.0 now? The best practice should be updated if that's the case... I can open an issue for that, depending on what's planned here.

**Comment by mlopatkin** [comment]
@gradle/bt-cortex let's try to at least understand how to move forward with it in 9.4 timeframe.
